### PR TITLE
Add composable-test-cases project to check JB compose compiler for all targets

### PR DESCRIPTION
**Pull request by eymar**
Please see a README.MD to get an idea of what this PR is trying to achieve.

### CODE REVIEW

**Comment by igordmn** [comment]
Thanks!

Shouldn't it be `androidx` repo? We need to enable github actions unit tests in the future.

If it is difficult to integrate with sources, we can just move it there, and call multiple gradle tasks

**Comment by eymar** [comment]
@igordmn 

We can place it in androidx repo. In that case we have such questions:

We use jb-main branch as a main development branch and we use release/compiler/a.b.c branches to build and publish the compose compiler plugin. 
Usually we keep only a few changes in `release/compiler/a.b.c`: it's based on a Jetpack Compose compiler release commit +   we add our patches for k/js, k/native. 
If we want to run those tests from `release/compiler/a.b.c` branch in github actions then it means we should keep more than just a few patches. 

+ Do we need to add these tests to jb-main branch? It would make sense, because jb-main is our main dev branch. But it would also mean that changes in the tests in jb-main have to be synced with `release/compiler/a.b.c`, which implies that we have 2 places to maintain. (same is true for patches, but their number is expected to be low because we upstream them. But we won't usptream the tests in this PR)
____

Keeping it outside of androidx (in compose-jb/master) seems to not add much issues, since in order to build and publish the compiler plugin we have to checkout the compose-jb too right now (at least it's how it works on our CI).
And if we checkout compose-jb, then we would have this tests checked out right away too. 

What do you think?

**Comment by igordmn** [comment]
> use release/compiler/a.b.c branches to build and publish the compose compiler plugin.

I didn't take it into account. In that case, indeed, it is simpler to have them here.

Let's finish this PR then.

**Comment by eymar** [comment]
@igordmn 
I've managed to make it not fail on CI with timeouts by passing the test coroutine context to the recomposer. 
Not this PR is ready for a final review.

File: compose/integrations/composable-test-cases/settings.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    module(":testcase-inheritance-funInterface-lib", "testcases/inheritance/funInterface/lib")
+    module(":testcase-inheritance-funInterface-main", "testcases/inheritance/funInterface/main")
+
+    module(":testcase-constructors-lib", "testcases/constructors/lib")
+    module(":testcase-constructors-main", "testcases/constructors/main")
+
+    module(":testcase-anonymousObjects-lib", "testcases/anonymousObjects/lib")
+    module(":testcase-anonymousObjects-main", "testcases/anonymousObjects/main")
+
+    module(":testcase-valueClass-lib", "testcases/valueClass/lib")
```
Comment (igordmn) [inline compose/integrations/composable-test-cases/settings.gradle.kts:116]: I have 2 failing tests on Windows:
```
> Task :testcase-valueClass-main:compileTestKotlinLinuxX64 FAILED
e: Compilation failed: Sequence is empty.

 * Source files: Tests.kt
 * Compiler version info: Konan: 1.8.10 / Kotlin: 1.8.10
 * Output kind: LIBRARY

e: java.util.NoSuchElementException: Sequence is empty.
        at kotlin.sequences.SequencesKt___SequencesKt.first(_Sequences.kt:112)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.defaultValue(ComposerParamTransformer.kt:261)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.defaultValue$default(ComposerParamTransformer.kt:240)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.defaultArgumentFor(ComposerParamTransformer.kt:224)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.withComposerParamIfNeeded(ComposerParamTransformer.kt:176)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer$copyWithComposerParam$2$3.visitCall(ComposerParamTransformer.kt:555)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitCall(IrElementTransformerVoid.kt:215)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitCall(IrElementTransformerVoid.kt:24)
...
        at org.jetbrains.kotlin.ir.declarations.IrFunction.transformChildren(IrFunction.kt:61)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoidKt.transformChildrenVoid(IrElementTransformerVoid.kt:346)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.copyWithComposerParam(ComposerParamTransformer.kt:508)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.withComposerParamIfNeeded(ComposerParamTransformer.kt:306)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.visitSimpleFunction(ComposerParamTransformer.kt:139)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitSimpleFunction(IrElementTransformerVoid.kt:73)


> Task :testcase-valueClass-main:compileTestKotlinDesktop
w:  `suppressKotlinVersionCompatibilityCheck` is set to the same version of Kotlin that the Compose Compiler was already expecting (Kotlin 1.8.10), and thus has no effect and should be removed.

> Task :testcase-valueClass-main:compileTestKotlinJs FAILED
e: java.util.NoSuchElementException: Sequence is empty.
        at kotlin.sequences.SequencesKt___SequencesKt.first(_Sequences.kt:112)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.defaultValue(ComposerParamTransformer.kt:261)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.defaultValue$default(ComposerParamTransformer.kt:240)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.defaultArgumentFor(ComposerParamTransformer.kt:224)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.withComposerParamIfNeeded(ComposerParamTransformer.kt:176)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer$copyWithComposerParam$2$3.visitCall(ComposerParamTransformer.kt:555)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitCall(IrElementTransformerVoid.kt:215)
...
        at org.jetbrains.kotlin.ir.declarations.IrFunction.transformChildren(IrFunction.kt:61)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoidKt.transformChildrenVoid(IrElementTransformerVoid.kt:346)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.copyWithComposerParam(ComposerParamTransformer.kt:508)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.withComposerParamIfNeeded(ComposerParamTransformer.kt:306)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.visitSimpleFunction(ComposerParamTransformer.kt:139)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitSimpleFunction(IrElementTransformerVoid.kt:73)
...
        at org.jetbrains.kotlin.ir.util.TransformKt.transformInPlace(transform.kt:35)
        at org.jetbrains.kotlin.ir.declarations.IrModuleFragment.transformChildren(IrModuleFragment.kt:51)
        at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoidKt.transformChildrenVoid(IrElementTransformerVoid.kt:346)
        at androidx.compose.compiler.plugins.kotlin.lower.ComposerParamTransformer.lower(ComposerParamTransformer.kt:105)
        at androidx.compose.compiler.plugins.kotlin.ComposeIrGenerationExtension.generate(ComposeIrGenerationExtension.kt:153)
        at org.jetbrains.kotlin.ir.backend.js.KlibKt.generateModuleFragmentWithPlugins$lambda$20(klib.kt:552)
        at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment(Psi2IrTranslator.kt:104)
```

Comment (eymar) [reply compose/integrations/composable-test-cases/settings.gradle.kts:116]: I forgot to update Readme.

Could you please try the latest version
compose.kotlinCompilerPluginVersion=1.4.2-rc03

Comment (igordmn) [reply compose/integrations/composable-test-cases/settings.gradle.kts:116]: It works now!

**Review by igordmn — APPROVED** [review]

File: compose/integrations/composable-test-cases/README.MD [hunk 2]
```diff
@@ -0,0 +1,53 @@
+## Run:
+
+`./gradlew build -Pcompose.kotlinCompilerPluginVersion=1.4.2-rc02 -Pkotlin.version=1.8.10`
```
Comment (eymar) [inline compose/integrations/composable-test-cases/README.MD:3]:
```suggestion
`./gradlew build -Pcompose.kotlinCompilerPluginVersion=1.4.2-rc03 -Pkotlin.version=1.8.10`
```

File: compose/integrations/composable-test-cases/README.MD [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+to build and run the tests,
+
+or
+`./gradlew allTests -Pcompose.kotlinCompilerPluginVersion=1.4.2-rc02 -Pkotlin.version=1.8.10`
```
Comment (eymar) [inline compose/integrations/composable-test-cases/README.MD:7]:
```suggestion
`./gradlew build -Pcompose.kotlinCompilerPluginVersion=1.4.2-rc03 -Pkotlin.version=1.8.10`
to build and run the tests,

or
`./gradlew allTests -Pcompose.kotlinCompilerPluginVersion=1.4.2-rc03 -Pkotlin.version=1.8.10`
```

**Pull request merged**